src/NCalc/Expression.cs

```diff
@@ -88,107 +89,75 @@ public static bool CacheEnabled
-                }
-                catch (ParseCanceledException)
-                {
-                    lexer.Reset();
+            LogicalExpression logicalExpression = null;
 
-                    parser.Reset();
-                    parser.ErrorHandler = new DefaultErrorStrategy();
-                    parser.Interpreter.PredictionMode = PredictionMode.LL;
-                    parser.AddErrorListener(parserErrorListener);
-                    logicalExpression = parser.ncalcExpression().value;
-                }
+            try
+            {
+                logicalExpression = parser.ncalcExpression().value;
+            }
+            catch (ParseCanceledException)
+            {
+                lexer.Reset();
 
-                if (parserErrorListener.Errors.Count > 0 || lexerErrorListener.Errors.Count > 0)
-                {
-                    var errors = string.Join(Environment.NewLine,
-                        lexerErrorListener.Errors.Select(e => e.ToString()).Union(
-                        parserErrorListener.Errors.Select(e => e.ToString()))
-                    );
+                parser.Reset();
+                parser.ErrorHandler = new DefaultErrorStrategy();
+                parser.Interpreter.PredictionMode = PredictionMode.LL;
+                parser.AddErrorListener(parserErrorListener);
+                logicalExpression = parser.ncalcExpression().value;
+            }
 
-                    throw new EvaluationException(errors);
-                }
+            if (parserErrorListener.Errors.Count > 0 || lexerErrorListener.Errors.Count > 0)
+            {
+                var errors = string.Join(Environment.NewLine,
+                    lexerErrorListener.Errors.Select(e => e.ToString()).Union(
+                    parserErrorListener.Errors.Select(e => e.ToString()))
+                );
 
-                if (_cacheEnabled && !nocache)
-                {
-                    try
-                    {
-                        if (_compiledExpressions.TryGetValue(expression, out var wr))
-                        {
-                            _compiledExpressions.TryUpdate(expression, new WeakReference(logicalExpression), wr);
-                        }
-                        else
-                        {
-                            _compiledExpressions.TryAdd(expression, new WeakReference(logicalExpression));
-                        }
-                    }
-                    finally
-                    {
-                        CleanCache();
-                    }
-                    //Debug.WriteLine("Expression added to cache: " + expression);
-                }
+                throw new EvaluationException(errors);
+            }
+
+            if (_cacheEnabled && !nocache)
+            {
+                _compiledExpressions[expression] = new WeakReference<LogicalExpression>(logicalExpression);
+
+                CleanCache();

```

Comment:
I'm not sure why it's important to prune the cache every time a new expression is compiled and could add a decent amount of overhead if someone has a high volume of different expressions.  This dictionary should be relatively light on memory usage since each item is a string and a weak reference.  

This seems like it should be part of a scheduled background job or potentially happen only after a certain number of compilations.

```
// no clue what the ideal number is
compilations++;
if (compilations % 1000 == 0)
{
  CleanCache();
}
```

Reply:
You're right, but for now I don't want a breaking change. The best way will be to add support for cache expiration

Reply:
Why would it be a breaking change?  This just delays pruning of the _compiledExpressions dictionary.  If a WeakReference is garbage collected it will still be updated on a re-Compile().

Reply:
You're right, it's not a breaking change, but still bot happy with your suggestion. What if we will remove from the cache on retrieving the value from dictionary when the value is missing (garbage collected)?

Reply:
I've never looked at CleanCache before this PR, but it seems to loop through every cached weak reference, looking for entries that have been garbage-collected; those that have been garbage-collected are removed from the cache.

> What if we will remove from the cache on retrieving the value from dictionary when the value is missing (garbage collected)?

I'm sorry, I'm not sure I understand the concern you're communicating.  Pruning the collected entries seems to be the intent.

Reply:
I've created another PR and removed the `CleanCache` method completely and cleaning the cache on retrive value. What do you think?

Reply:
We can also look at [ConditionalWeakTable](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2)

Reply:
I wasn't aware of ConditionalWeakTable, but based on the documentation I don't think it can be used in this scenario.  

From the docs:

> [ConditionalWeakTable<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-8.0) automatically removes the key/value entry as soon as no other references to a key exist outside the table. 

NCalc wants a weak reference to the compiled LogicalExpression, not the string expression.  ConditionalWeakTable removes entries when there are no longer references to the key, not the value.  I did some searching, but there doesn't seem to be an equivalent table in the BCL that works off the values.  I'd hoped you could define a custom key class that wraps the string and LogicalExpression, but unfortunately the docs indicate that ConditionalWeakTable doesn't call GetHashCode() and only does reference equality.

Reply:
Sorry for side tracking this so much.  I'd say put the PR back to the way you originally had it.  Use WeakReference<LogicalExpression> and TryGetTarget().  I'll make a separate PR for optimizing the cleanup.

Reply:
No problem, I've reverted the last 2 commits